### PR TITLE
Enhance Microsoft Word Comment Command: Press Twice to Present Comment Content in Browsable message.

### LINKED ISSUE #14628: Feature request: 'Reports the text of the comment where the System caret is located.' supports opening in a browsable dialog
### Is your feature request related to a problem? Please describe.
Sometimes the text of comment is too long, in order to understand, we may need to navigate by sentence or by word.
### Describe the solution you'd like
Press NVDA +Alt +C twice to display a browsable dialog to contain the text of comment.
### Describe alternatives you've considered
None
### Additional context
I would like to open a PR myself to contribute to this.

### PR DESCRIPTION
**Pull request by cary-rowen**
### Link to issue number:
Partially fixed #14628 
Replaces abandoned #15987

**Thanks to @tseykovets for most of the work.**

### Summary of the issue:
Sometimes the text of comment is too long, in order to understand, we may need to navigate by sentence or by word.
### Description of user facing changes
This pull request adds the ability to present the text of a comment in browse mode by pressing command twice (NVDA+Alt+C).
### Description of development approach
A handler for the number of presses has been added to the script (if repeats).
If pressed twice, presents the information in browse mode (ui.browseableMessage).
### Testing strategy:
Tested manually in Microsoft Word.
### Known issues with pull request:
Unknown
### Code Review Checklist:

- [X] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [X] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] API is compatible with existing add-ons.
- [ ] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Enhanced script to report text at the system caret location and provide additional information when pressed twice in browse mode.
  - Improved message handling in comments, presenting different messages based on script repeats.

### CODE REVIEW

File: user_docs/en/changes.md [hunk 1]
```diff
@@ -6,6 +6,9 @@
 
 ### New Features
 
+* Microsoft Office:
+  * Enhance Microsoft Word Comment Command: Press Twice to Present Comment Content in Browsable message. (#16800, @Cary-Rowen)
```
Comment (seanbudd) [inline user_docs/en/changes.md:10]:
```suggestion
* Enhance Microsoft Word comment command: press twice to present comment content in browsable message. (#16800, @Cary-Rowen)
```

File: source/NVDAObjects/UIA/wordDocument.py [hunk 2]
```diff
@@ -611,15 +612,27 @@ def _caretMoveBySentenceHelper(self, gesture, direction):
...
 		caretInfo=self.makeTextInfo(textInfos.POSITION_CARET)
 		commentInfo = getCommentInfoFromPosition(caretInfo)
 		if commentInfo is not None:
-			ui.message(getPresentableCommentInfoFromPosition(commentInfo))
+			text = getPresentableCommentInfoFromPosition(commentInfo)
+			repeats = scriptHandler.getLastScriptRepeatCount()
+			if repeats == 0:
+				ui.message(text)
+			elif repeats == 1:
+				ui.browseableMessage(text,
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/wordDocument.py:632]:
```suggestion
				ui.browseableMessage(
					text,
```

File: source/NVDAObjects/UIA/wordDocument.py [hunk 3]
```diff
@@ -611,15 +612,27 @@ def _caretMoveBySentenceHelper(self, gesture, direction):
 	@script(
 		gesture="kb:NVDA+alt+c",
 		# Translators: a description for a script that reports the comment at the caret.
-		description=_("Reports the text of the comment where the system caret is located."),
+		description=_(
+			# Translators: a description for a script that reports the comment at the caret.
+			"Reports the text of the comment where the system caret is located."
+			" If pressed twice, presents the information in browse mode"
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/wordDocument.py:618]:
```suggestion
			" If pressed twice, presents the information in a browsable message"
```

File: source/NVDAObjects/UIA/wordDocument.py [hunk 4]
```diff
@@ -611,15 +612,27 @@ def _caretMoveBySentenceHelper(self, gesture, direction):
...
+		),
 		category=SCRCAT_SYSTEMCARET,
 		speakOnDemand=True,
 	)
 	def script_reportCurrentComment(self,gesture):
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/wordDocument.py:623]: please lint and add a type hint for gesture

File: user_docs/en/userGuide.md [hunk 5]
```diff
@@ -1488,6 +1488,7 @@ The Elements List can list headings, links, annotations (which includes comments
 
 <!-- KC:beginInclude -->
 To report any comments at the current caret position, press NVDA+alt+c.
```
Comment (seanbudd) [inline user_docs/en/userGuide.md:1490]:
```suggestion
To report any comments at the current caret position, press `NVDA+alt+c`.
```

File: user_docs/en/userGuide.md [hunk 6]
```diff
@@ -1488,6 +1488,7 @@ The Elements List can list headings, links, annotations (which includes comments
...
+Pressing twice shows the information in a window.
```
Comment (seanbudd) [inline user_docs/en/userGuide.md:1491]:
```suggestion
Pressing twice shows the information in a browsable message.
```

**Review by seanbudd — CHANGES_REQUESTED** [review]

File: user_docs/en/changes.md [hunk 7]
```diff
@@ -6,8 +6,7 @@
 
 ### New Features
 
-* Microsoft Office:
-  * Enhance Microsoft Word Comment Command: Press Twice to Present Comment Content in Browsable message. (#16800, @Cary-Rowen)
+* Enhance Microsoft Word comment command: press twice to present comment content in browsable message. (#16800, @Cary-Rowen)
```
Comment (seanbudd) [inline user_docs/en/changes.md:9]:
```suggestion
* Enhanced Microsoft Word comment command: press twice to present comment content in browsable message. (#16800, @Cary-Rowen)
```

**Review by seanbudd — APPROVED** [review]

**Comment by seanbudd** [comment]
Thanks @cary-rowen

**Comment by seanbudd** [comment]
@cary-rowen can you confirm you've tested the latest changes you've pushed?

**Comment by CyrilleB79** [comment]
I wonder if we cannot propose another UX here, e.g.:
* First press reports the comment
* Second press jumps in the comment area

This would allow to navigate in the comment, but also in the answers of the comment.

Also, it's worth noting that there are various issues related to Word comments (#14985, #9685, #4336). The use case of comments with replies is much less common than single comments. But should we really continue to ignore it totally in the UX.

Of course, a third path can be to merge this PR as is and to modify again the UX when the other issues are handled, with the risk though that users used to the new UX provided in this PR (the browseable message) be disappointed if it is removed in the future.

It should be NV Access decision (@seanbudd, @Qchristensen), but I just write this here to have all the related issues in mind for everyone.

**Comment by cary-rowen** [comment]
cc @seanbudd 
> can you confirm you've tested the latest changes you've pushed?
Yes I tested locally using Microsoft 365MSO (Version 2406 Build 16.0.17726.20078) 64 Bit.

**Comment by cary-rowen** [comment]
Hi @CyrilleB79 

> First press reports the comment
> Second press jumps in the comment area

I just to confirm, by "comment area" you mentioned here, do you mean "sidetrack" which can be focused with Alt+F12?

> Also, it's worth noting that there are various issues related to Word comments 

Yes, I'm aware of these reported issues as well, especially #14985 which deserves to be fixed in a separate PR.

Regarding the UX considerations you mentioned, I'd also like to hear NV Access's thoughts. Especially if Alt+F12 can focus the sidetrack.

**Comment by CyrilleB79** [comment]
> I just to confirm, by "comment area" you mentioned here, do you mean "sidetrack" which can be focused with Alt+F12?

On my end, `alt+F12` is not working. Though, there is an important difference between [modern comments](https://support.microsoft.com/en-us/office/using-modern-comments-8d3f868a-867e-4df2-8c68-bf96671641e2) and classical ones. Modern ones seem to be available only with Word 365; I have Word 2016 here.

**Comment by cary-rowen** [comment]
鉴于这些Office 版本之间的差异，我不知道这个 PR 能做些什么。

实际上，这个  PR 的目标很明确，而关于评论相关的其他问题

**Comment by cary-rowen** [comment]
Hi @CyrilleB79 

Due to the differences between Office versions, I'm not quite sure what we can do in this PR to unify the UX.

Whether it is Office365 or 2016, Office seems to have native shortcut keys for focusing on the comment area.

NVDA users will not be unfamiliar with browsable dialog. For NVDA, such browsable message are used very widely, such as report formatting information.
As for other issues related to report comments, they are definitely worth fixing in separate PRs.

@seanbudd Any idea?
Thanks

**Comment by seanbudd** [comment]
we are happy to accept this as-is. In the future we can consider performing alt+f12 on triple press

**Comment by CyrilleB79** [comment]
@cary-rowen wrote:
> Due to the differences between Office versions, I'm not quite sure what we can do in this PR to unify the UX.
> 
> Whether it is Office365 or 2016, Office seems to have native shortcut keys for focusing on the comment area.

Actually, I didn't know that there were a native shortcut to edit a comment in Word 365. I do not know of an equivalent shortcut in Word 2016, except of course the sequential one corresponding to what is found in the ribbon.

@cary-rowen also wrote:
> NVDA users will not be unfamiliar with browsable dialog. For NVDA, such browsable message are used very widely, such as report formatting information. As for other issues related to report comments, they are definitely worth fixing in separate PRs.

I agree that the browseable message is very familiar for NVDA users and is probably easier to navigate than Word's native comment area.
Since I am not able to suggest something to improve the UX in Word's native comment area, nor to suggest a way to reach a unified experience between 365 and 2016, you can just ignore my previous suggestion.

If possible, a second step (in a separate PR) could be to add the replies in the browseable message, in HTML, with separators between each one.

**Comment by seanbudd** [comment]
Can you please fix the lint @cary-rowen - see the build artifact [licenseCheckResults.md](https://ci.appveyor.com/api/buildjobs/30sbsris4jeyr6xh/artifacts/licenseCheckResults.md)

**Comment by seanbudd** [comment]
Please disregard previous comment

Can you please fix the lint @cary-rowen - I'm not sure what's wrong with the lint artifacts, can you run `runlint.bat` locally and report the results?

**Comment by cary-rowen** [comment]
Hi @seanbudd 
I executed `runlint.bat` and the following information was output. Can I commit these changes?

```
D:\git\nvda>runlint.bat
Ensuring NVDA Python virtual environment
call ruff check --fix
All checks passed!
Deactivating NVDA Python virtual environment
Ensuring NVDA Python virtual environment
call ruff format
557 files reformatted, 15 files left unchanged
Deactivating NVDA Python virtual environment
```

**Comment by cary-rowen** [comment]
Hi @CyrilleB79 

Thank you for your previous thoughts in bringing these existing issues to the attention of the community.

> If possible, a second step (in a separate PR) could be to add the replies in the browseable message, in HTML, with separators between each one.

This is a good idea. I find it very useful.

**Comment by seanbudd** [comment]
@cary-rowen - is that after merging the latest master? There really shouldn't be that many changes.

**Comment by seanbudd** [comment]
@cary-rowen - see the lint results here: https://ci.appveyor.com/project/NVAccess/nvda/builds/50189139#L5378
please add the trailing commas

**Comment by hwf1324** [comment]
> @cary-rowen - is that after merging the latest master? There really shouldn't be that many changes.

For Git for Windows, it seems that even if you merge the latest master, the local files don't update to the end of the LF line. It also doesn't seem to create new commits after committing those changes though?

**Review by seanbudd — APPROVED** [review]

**Pull request merged**